### ISSUE TITLE
Discussion: Manager for Unified Devices

### ISSUE THREAD

**Issue by jramapuram** [arrayfire/arrayfire#1537]
So I'm been building [hal](https://github.com/jramapuram/hal) and have had to glue together some code to manage arrays across opencl/cuda/cpu boundaries. I use arrayfire-cpu for loading my data after which I use the code below to do a GPU copy [ML requires these constant swapping operations as we deal with minibatches]. Sure I could just use native float's and such, but that loses the power of af! An ideal situation is having the same API across devices (and do complicated ops like SVD, etc). My solution is somewhat mediocre: I basically lock and swap backends:

```rust
 pub fn swap_device(&self, device: Device)
  {
    let mut c = self.current.lock().unwrap();
    if c.backend != device.backend || c.id != device.id
    {
      assert!(self.devices.contains(&device)
              , "device backend = {} | available = {:?}"
              , device.backend, self.devices);
      // println!("Swapping {}/{} to {}/{}", c.backend, c.id
      //          , device.backend, device.id);
      set_device(device);
      *c = device;
    }
  }

pub fn swap_array_backend<T>(&self, input: &Array
                               , input_device: Device
                               , target_device: Device) -> Array
    where T: HasAfEnum + Zero + Clone
  {
    // return if the devices match
    if input_device.id == target_device.id
        && input_device.backend == target_device.backend
    {
      return input.clone() // increases the ref count
    }

    // we have done something bad if the following triggers
    let ib = input.get_backend();
    if input_device.backend != Backend::DEFAULT {
      assert!(ib == input_device.backend
              , "provided src was {:?}, but actually {:?}"
              , input_device.backend
              , ib);
    }

    // ensure we are on the old device
    self.swap_device(input_device);

    // copy data to the host
    let dims = input.dims();
    let mut buffer: Vec<T> = vec![T::zero(); dims.elements() as usize];
    input.host(&mut buffer);

    // swap to the new device
    self.swap_device(target_device);
    Array::new::<T>(&buffer, dims)
  }
```

I notice random errors [eg: weak array references, etc] on occasion when i forget to either move to the correct backend or when I do operations between arrays of different backends. This has let me to believe that there must be a better solution to all of this & want to start a discussion on this topic. 

At first glance I would imagine running a thread per backend (and per device) and auto sorting ops from each backend to it's respective thread. This is "ok", but there is no way to do this currently in af (correct me if I'm wrong here) as we would in fact need multiple processes. This also runs into the issue of when you have an op that transcends compute devices (eg: array on CPU & array on CUDA). What would be the strategy for these? Maybe we could intelligently "upcast" device types? I.e. the user picks a strategy for ops (eg: `enum UpcastStrategy::TryAutoUpCast` or `UpcastStrategy::ErrorOut` , etc. 

Thoughts @9prady9 , @pavanky , @shehzan10  ?

**Comment by pavanky** [comment]
Bringing it to @umar456's attention as well.

@jramapuram This is a good solution but is not easy to implement *properly*. Thanks for starting the conversation. I opened a related issue sometime ago: https://github.com/arrayfire/arrayfire/issues/1268

May be we can have a solution that solves both problems.

**Comment by jramapuram** [comment]
@pavanky : Definitely. I think the two are highly correlated. It will definitely be a non-trivial solution, but I think a thread per context makes the most sense in my mind. There would be a sort of 'manager' thread that enqueue's data & engages a conditional wait for each 'backend' thread. So, standard event loop. I think the problem as I mentioned is that currently we need a new **process** per af backend. I think a good first start would be to make that accessible within the same process scope.
